felix/bpf-gpl/tc.c

```diff
@@ -1226,6 +1291,39 @@ static CALI_BPF_INLINE struct fwd calico_tc_skb_accepted(struct cali_tc_ctx *ctx
 		fib = false; /* Disable FIB because we want to drop to iptables */
 	}
 
+	if (CALI_F_TO_HEP && !skb_seen(skb) && is_dnat) {
+		struct cali_rt *r = cali_rt_lookup(state->post_nat_ip_dst);
+
+		if (r) {
+			if (cali_rt_flags_local_workload(r->flags)) {
+				state->ct_result.ifindex_fwd = r->if_index;
+				CALI_DEBUG("NP local WL %x:%d on HEP\n",
+						bpf_htonl(state->post_nat_ip_dst), state->post_nat_dport);
+				ctx->state->flags |= CALI_ST_CT_NP_LOOP;
+				fib = true; /* Enforce FIB since we want to redirect */
+			} else if (cali_rt_flags_remote_workload(r->flags)) {
+				if (CALI_F_LO || CALI_F_MAIN) {
+					state->ct_result.ifindex_fwd = NATIN_IFACE  ;
+					CALI_DEBUG("NP remote WL %x:%d on LO or main HEP\n",
+						bpf_htonl(state->post_nat_ip_dst), state->post_nat_dport);
+					ctx->state->flags |= CALI_ST_CT_NP_LOOP;
+				}
+				ctx->state->flags |= CALI_ST_CT_NP_REMOTE;
+				fib = true; /* Enforce FIB since we want to redirect */
+			}
+		} else {
+			/* Did not find a route for a service BE? Should be just
```

Comment:
What about out-of-cluster service backends such as the k8s API server?  I don't think they'll have routes